### DISCUSSION TITLE
Issue with Switch Position and Alarm Condition Reset in Genmon

### DISCUSSION THREAD

**Discussion by ericsmith66** [jgyates/genmon#1133] [Questions]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).

### Version

Genmon
Version V1.19.03

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have submitted logs

### What is/are your question(s)?

Hello,
I’ve been using Genmon on my Generac Liquid-cooled 48kW generator (model: RG04845ANAX) since mid-May and have found the software to be excellent. I’m currently developing a HomeKit plugin to link Genmon with span.io breaker panels.
I’ve successfully tested basic functions like starting and stopping the generator via the JSON API. However, I’ve encountered an issue when trying to use the Switch Position and Alarm Condition reset functions.
When I click on these buttons in the UI, I receive a 200 response code, indicating that the request was successful. However, the buttons do not appear as “selected” in the UI, and there is no corresponding change on the physical generator panel.
Additionally, when I originally installed Genmon, it did not detect the model number or the serial number of the generator. I have not yet tested the Start Generator and Transfer Switch functions (I need to get permission from my wife first).
I’ve checked the logs but haven’t found anything related to these specific button pushes. I have submitted the logs and the registers via the about page.
Any guidance on how to resolve this issue would be greatly appreciated.

Thanks 
Eric Smith

**Comment by jgyates** [comment]
To clarify, you issue the remote command for Auto,Manual, Off from the UI and the generator does not respond, correct. Also based on your logs you have a liquid cooled 4.5L model (which may be relevant).

As a test, with genmon running run this command from the genmon folder on the pi.

<img width="439" alt="Screenshot 2024-06-04 at 3 46 32 PM" src="https://github.com/jgyates/genmon/assets/4513286/507ecc6c-03c0-4092-900b-bbe387e2d201">

This should set the remote Off command for the switch, then back to Auto. Does genmon update the UI based on this test (off then back to auto) and do you get an email notice that the state of the generator changed twice? 

For now let's not worry about the state of the button in the UI as that is separate from the actual state of the controller.

NOTE: the Off, Auto, Manual commands only work on liquid cooled controllers. I assume the also work on the newer 4.5L liquid cooled models as well. The 4.5L controllers models have a different build of the firmware. I don't have a lot of data points on the 4.5L models so there is a possibility that that functionality many not work on this unit.

**Comment by ericsmith66** [comment]
I appreciate your prompt response. I conducted the test as you suggested and here are the results:
I executed the following commands in the Genmon directory on my Raspberry Pi:

ericsmith66@genmonpi4:~/genmon $ python3 ClientInterface.py
OK : Auto, Off - Ready
>setremote=off
Invalid GENERATOR command: all commands must be prefixed by "generator: "
>generator: setremote=off
Remote command sent successfully
>generator: setremote=auto 
Remote command sent successfully

After running these commands, I observed the generator’s state. Unfortunately, it did not switch to the ‘off’ state as expected.
Additionally, I noticed several error messages in the mymodbus.log file. Here are some of the error messages:

2024-06-04 16:42:07,379 : Modbus Exception: Acknowledge: 05 , Modbus Command: 90 , Sequence: 1337516
2024-06-04 16:42:07,379 : Error Receiving slave packet for register 0003
...
2024-06-04 16:46:09,749 : Modbus Exception: Illegal Data Value: 03 , Modbus Command: 90 , Sequence: 1360625
2024-06-04 16:46:09,749 : Error Receiving slave packet for register ea60

For additional context, my generator was purchased new in August of last year. The control panel appears to be an Evolution 1.0 (it does not have a 2.0 login). I am using a pHAT v2.6+ mounted on a Raspberry Pi 4 for this setup.

I also noticed that the following registers are red: Utility Voltage
(Holding:0009)
HEX: 00f6
DEC: 246 | HI:LO: 0:0
Battery Voltage
(Holding:000a)

HEX: 0086
DEC: 134 | HI:LO: 0:86
Engine Run Hours (Nexus, EvoAC)
(Holding:000b)

Generator Time HH:MM
(Holding:000e)
HEX: 1107
DEC: 4359 | HI:LO: 4:83


Battery Voltage
(Holding:000a)
HEX: 0086
DEC: 134 | HI:LO: 0:86

Battery Charger CT (EvoLC)
(Holding:05ee)
HEX: 018e
DEC: 398 | HI:LO: 0:18

**Reply by jgyates** [reply]
I looks like your serial data rate is set to 115200 as per the [known issues wiki entry](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues#18-liquid-cooled-45l-model-serial-data-rate). One thing to try is to set the data rate back to 9600 and see if that makes a difference.

Another thing that appears to be happening is that the Evo2.0 decryption code is getting activated (due to the modbus exceptions). Genmon will attempt to use the decrption / encapsulation methods if more than a given number of modbus exceptions occur (the never occur on non 4.5L Evo1.0 controllers. I have only seen them on Evo 2.0.

How often are modbus exceptions occurring? If you restart genmon do the modbus exceptions stop?

**Reply by jgyates** [reply]
Just a theory: if the remote command is issued and it generates a modbus exception because it is not supported or not formatted correctly that could be triggering the Evo2 code. Presently the Evo2 code is activated when 2 modbus exceptions are detected. This can be changed with the addition of and entry in the /etc/genmon/genmon.conf file. If you add this line to /etc/genmon/genmon.conf


     evo2_max_except=10

Then the evo 2 code will activate after 10 modbus exceptions.

**Reply by jgyates** [reply]
once we figure out what is happening with the remote commands and modbus exceptions I can make any changes to the 4.5L code to prevent the Evo2 code from being activated (if that is what is causing the issues).

**Reply by ericsmith66** [reply]
Taking the serial data rate down, causes all data to stop receiving by genmon. I just rebooted and have not seen a new error in modes ( for the last 4 minutes ). I could not see a pattern in the earlier logs.

**Reply by jgyates** [reply]
ok, that is good. Try the remote commands again and see if they generate a modbus exception. After 2 modbus exceptions the decryption code will kick in and it is likely nothing will work after that.

**Reply by jgyates** [reply]
If we can isolate which remote commands work and which ones do not I can update the user interface to not show those options for 4.5L models.

**Reply by ericsmith66** [reply]
Ok I tried to turn the generator off with the same result. I saw a couple of errors in the modbus log.  I then did a start generator, and that was successful. I then did a stop and that also worked. I just upped the evo2 max exceptions . I will reboot and retest.

**Reply by jgyates** [reply]
I am getting an automated "unknown engine state" email from you also that shows the engine status registers 0001 value of 000f0000. I assume this occurs when you issue the off remote command. Can you conform that? It may be that the behavior of the 4.5L is a little different with remote commands. It may be useful to issue the remote switch off command, look at the contents of registers 0001 then look at the controller LCD to see the status from the controller. This should help clear things up.

**Reply by ericsmith66** [reply]
1. Ok - Reboot with the max  evo2_max_except=10. Then I did the follow
2. Executed the Off button from the UI - no effect and then I saw errors in the modbus log (2) 
3. Executed the Manual button from the UI - no effect and then saw errors in the modbus log
4. executed the reset alarm button - no effect but no errors in the log.
I am not sure which of these events triggered the email . If you have a sequence you would like me to go thru I can try and get you a proper test result. ( do you have a list off all the remote commands ?)

**Reply by ericsmith66** [reply]
Just did a reboot , I will stop making any changes until you tell me what you would like to test.

**Reply by jgyates** [reply]
ok. lets try this:

1) add this entry to /etc/genmon/genmon.conf

         evo2_bypass_encapsulation=True

save the file. Then:

2) Update to the latest from the about page (version number has not changed so it is still 1.19.03)
 
Once the update is applied lets try the remote off and auto commands. I still expect some modbus exceptions but try it in this order, noting any changes in modbus exceptions at each command:

  remote switch Off

Are there mod bus exceptions? What does the UI state, what does the controller LDC state?

  remote switch Auto

Are there mod bus exceptions? What does the UI state, what does the controller LDC state?

  remote switch Manual

Are there mod bus exceptions? What does the UI state, what does the controller LDC state? Did the generator start in manual mode

  remote switch Off

Are there mod bus exceptions? What does the UI state, what does the controller LDC state?

**Reply by jgyates** [reply]
I added code to handle what I think is the code 000f0000. I also added the new config option mentioned above to disable the Evo2 encapsulation. Once we nail down the issue with the modbus exceptions then that entry should not be needed but I thought it might make the testing / debugging easier.

**Reply by ericsmith66** [reply]
Ok starting now

**Reply by ericsmith66** [reply]
Here are my observation so far:

1. Modbus Error: Regardless of the button pressed (Manual, Off, or Auto), a Modbus error is consistently generated. Here are the error sample errors for each button:

    - Off Button: 
        - Time: 2024-06-04 18:41:24,926
        - Modbus Exception: Acknowledge: 05, Modbus Command: 90, Sequence: 219357
        - Error: Receiving slave packet for register 000

    - Auto Button (pressed when the generator is off):
        - Time: 2024-06-04 18:48:23,994
        - Modbus Exception: Acknowledge: 05, Modbus Command: 90, Sequence: 260281
        - Error: Receiving slave packet for register 0003

    - Manual Button:
        - Time: 2024-06-04 19:33:10,565
        - Modbus Exception: Acknowledge: 05, Modbus Command: 90, Sequence: 83376
        - Error: Receiving slave packet for register 0003

2. LCD State: The LCD state does not change regardless of the button pressed in the UI. The UI button state only changes if I manually change the switch state on the LCD.

3. Engine State: I received an Engine State of UNKNOWN: 000f0007 once, Clicking OFF does not always result in UNKNOWN: 000f0007. Unfortunately, I do not have a reliable method to reproduce.


I will keep trying to reproduce the 000f0007. Let me know if I can send you anything.

**Reply by jgyates** [reply]
This behavior leads me to believe that the remote Auto, Manual, Off commands are not supported by the 4.5L liquid cooled models. With more recent controller version Generac responds with an Acknowledge modbus exception (which means I got your command, but I can't do that) for commands that are not supported. 

I am going to disable this functionality in the UI. Note that if you issue the remote command via the command line tool or via a web API genmon will still send the command to the controller. If your firmware is ever updated and you want to try it again let me know.

I have checked in an update that disable this display of the Auto, Manual, Off buttons from the UI. 

Let me know if any other remote commands cause modbus exceptions.

**Reply by ericsmith66** [reply]
thanks for your help with this . i need to check and see if mobile link supports them .on this generator . i have mobile link in a box some place. if it does maybe i can sniff out what its doing.

**Reply by jgyates** [reply]
that would be great.

**Reply by ericsmith66** [reply]
Just a follow up. I tested the start/transfer switch and it worked without issue. I did notice that the transfer switch also put the generator into an unknown state. I was wondering if there is a way to tell what version of the firmware is installed. Generac just says call them / a service center .   
Thanks 
Eric Smith

**Reply by jgyates** [reply]
The firmware version is displayed on the Maintenance page in the Service section

**Reply by ericsmith66** [reply]
Just confirming on your modifications . The manual , auto and off commands are no longer showing up in the ui (as you discussed). the api's are still functional.
Regards 
Eric

**Reply by ericsmith66** [reply]
One last question. I am trying to confirm it the reset alarm state will work with my version of the firmware. Is there a safe way to put the generator in an alarm state , so that I can then try and clear it ?

**Reply by jgyates** [reply]
often times things like the oil, coolant level or tank bladder (for diesel systems) sensor will be connected to a sensor that you can disconnect / unclip at the cable. If you do this when the generator is off it may generate an alarm. I have done this in the past with a liquid cooled diesel to force alarms (not with a 4.5L however).
